front/components/assistant/conversation/actions/VisualizationIframeContent.tsx

```diff
@@ -0,0 +1,250 @@
+            type: contentType || undefined,
+          })
+        )
+      )
+      .catch(console.error);
+  }, [fileId, workspaceId]);
+  if (!fileId) {
+    return null;
+  }
+
+  console.log("returned data", data);
+  return data;
+}
+
+function makeIframeMessagePassingFunction<Params, Answer>(
+  methodName: RPCMethod,
+  actionId: number
+) {
+  return (params?: Params) => {
+    return new Promise<Answer>((resolve, reject) => {
+      console.log("sending a message to parent", params);
+      const messageUniqueId = Math.random().toString();
+      const listener = (event: MessageEvent) => {
+        if (event.data.messageUniqueId === messageUniqueId) {
+          if (event.data.error) {
+            reject(event.data.error);
+          } else {
+            resolve(event.data.result);
+          }
+          window.removeEventListener("message", listener);
+        }
+      };
+      window.addEventListener("message", listener);
+      window.top?.postMessage(
+        {
+          command: methodName,
+          messageUniqueId,
+          actionId,
+          params,
+        } satisfies CrossWindowRequest,
+        "*"
+      );
+    });
+  };
+}
+
+export function VisualizationIframe({
+  actionId,
+  workspaceId,
+}: {
+  actionId: string;
+  workspaceId: string;
+}) {
+  const [code, setCode] = useState<string | null>(null);
+  const [errored, setErrored] = useState<Error | null>(null);
+  const useFileWorkspaceWrapped = (fileId: string) =>
+    useFile(workspaceId, fileId);
+
+  useEffect(() => {
+    // get the code to execute
+    const getCodeToExecute = makeIframeMessagePassingFunction<
+      { actionId: string },
+      { code: string }
+    >("getCodeToExecute", parseInt(actionId, 10));
+    getCodeToExecute({ actionId })
+      .then((result) => {
+        const regex = /<visualization[^>]*>\s*([\s\S]*?)\s*<\/visualization>/;
+        let extractedCode: string | null = null;
+        const match = result.code.match(regex);
+        if (match && match[1]) {
+          extractedCode = match[1];
+          setCode(extractedCode);
+          console.log("got code to execute", result, extractedCode);
+        } else {
+          setErrored(new Error("No visualization code found"));
+        }
+      })
+      .catch(console.error);
+  }, [actionId]);
+
+  const retry = useMemo(() => {
+    return makeIframeMessagePassingFunction("retry", parseInt(actionId, 10));
+  }, [actionId]);
+
+  if (errored) {
+    return <VisualizationError error={errored} retry={() => retry()} />;
+  }
+  if (!code) {
+    return <Spinner variant="color" size="xxl" />;
+  }
+
+  const generatedCodeScope = {
+    recharts: rechartsAll,
+    react: reactAll,
+    papaparse: papaparseAll,
+    "@dust/react-hooks": { useFile: useFileWorkspaceWrapped },
+  };
+
+  const scope = {
+    import: {
+      recharts: rechartsAll,
+      react: reactAll,
+
+      "./local-file": importCode(code, { import: generatedCodeScope }),
```

Comment:
OOC why do we need a file at all ?

Reply:
This was just a copy / paste from the documentation from react-runner. I changed it to `@dust/generated-code` for the next commit 👍